### PR TITLE
Migrate map editor code to mettamap (#1117)

### PR DESCRIPTION
**Pull request by daveey**
Migrate map editor code to mettamap (#1117)

- integrated/reimplemented map editor in mettamap, with React and a more
powerful map viewer component
- added `mettamap/start.py` script that starts both frontend and
backend, for easier use (this is temporary until we merge mettamap and
observatory code)
- added a lot of related changes, fixes and improvements to MapViewer
component, including new sprites, compatibility with the ascii encoding
#850, etc.

There are a lot of obvious improvements that can be done here, like
undo/redo, copy/paste areas, etc. But this is the minimal checkpoint
that allowed me to replace `map_editor_ui` (and make it compatible with
new ascii encoding, which was the initial goal).

Note: I didn't want to reimplement wall sprite connection logic from
mettascope, so I opted to use the simplest single sprite, for now (with
gray background for emphasis).

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210624037525217)

Default border width=5 in mapgen (#1146)

Previously was set to 1, which was a mistake because we don't want
agents to see beyond map borders.

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210635438865246)

resolve device to cpu if set to cuda but no cuda available (#1177)

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210645927515704)

Grad stats (#1175)

Calculate grad mean, variance, and norm then plot to WandB depending on
`grad_mean_variance_interval` (default to 0 ie off).

This PR has admittedly less motivation than one with a direct need.
Rather, it tracks something that _may_ be useful in examining policy
collapse but remains to be tested. One potentially benefit (of a few
promised) would be to understand when training has actually plateaued:
not when reward seems flat but when grad variance starts to decrease.

**Theoretical Motivation**
Proposed in [Opening the black box of Deep Neural Networks via
Information](https://arxiv.org/pdf/1703.00810), models go through two
phases: fitting (drift) and compression (diffusion). During drift, the
model quickly fits and its loss curve approaches a plateau. During
diffusion, the model compresses its representation even though the loss
curve may not change significantly.

The gradient's mean is much higher than the variance during drift.
During diffusion, the relationship flips. And while the low mean, high
variance, phase seems very noisy, the noise actually acts as a
regularizer.

Thanks to Matt Bull for the recommendation here!

**In Practice**
See [this training
example](https://wandb.ai/metta-research/metta/runs/m_alexv_grad_stats_01)
and the "grad" plots. Training on the simple environment shows tiny mean
and variance the correlations just aren't there. Given the strength of
the theory, this is something would be interesting to investigate in
other training environments and as we start to make changes to the
simple environment.

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210644969846468)

Replace mine and generator with *_red (#1179)

Stop supporting auto converting from mine and generator to mine_red and
generator_red. This removes ambiguity and special casing, which will
make it easier to remove these types from code.

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210646337233238)

---------

Co-authored-by: S. Alex Smith <sasmith@stem.ai>

More Demo Mode work. (#1171)

* Make zooming more tighter.
* Fix odd display resizing bugs.
* Inventory on top of selection.

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210651984414358)

Use server for dashboard (#923)

Update dashboard FE to use the webapp apis to load data.

Will not actually merge this until we figure out deployment

---------

Co-authored-by: Paul Tsier <paul@elementanalytics.com>

Clean up policy names for training runs in observatory (#1185)

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210653065999174)

Co-authored-by: Paul Tsier <paul@elementanalytics.com>

fix: save individual pr analysis (#1108)

This updates the weekly PR summary script to save individual PR analyses
as separate text files alongside the main summary outputs.

## Changes

- Added `save_pr_summary_file()` function to save each PR summary as a
plain text file (`pr_XXXX.txt`)
- Updated console output to reflect the new file structure

## Output Structure

- `pr_digest_output.json` - Original PR data (input)
- `pr_summary_output.txt` - Discord-formatted summary
- `pr_1003.txt`, `pr_1004.txt`, etc. - Individual PR summaries

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210622413833012)

Add a feature spec (#1192)

Added a new `feature_spec()` method to MettaGrid that provides
comprehensive feature information including normalization values and
IDs.

This should allow models to adapt to how the environment communicates
features.

[Asana
Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210654248905742)

Co-authored-by: S. Alex Smith <sasmith@stem.ai>

cp

### CODE REVIEW

File: metta.code-workspace [hunk 1]
```diff
@@ -103,21 +109,24 @@
...
 			"streambuf": "cpp",
-			"string": "cpp",
 			"string_view": "cpp",
+			"string": "cpp",
+			"typeindex": "cpp",
 			"typeinfo": "cpp",
 			"unordered_map": "cpp",
 			"unordered_set": "cpp",
-			"valarray": "cpp",
+			"valarray": "cpp""valarray": "cpp",
```
Comment (sasmith) [inline metta.code-workspace:129]: broken

File: mettagrid/src/metta/mettagrid/mettagrid_c.cpp [hunk 2]
```diff
@@ -348,11 +353,16 @@ void MettaGrid::_step(py::array_t<ActionType, py::array::c_style> actions) {
...
+  std::vector<size_t> agent_indices(_agents.size());
+  std::iota(agent_indices.begin(), agent_indices.end(), 0);
+  std::shuffle(agent_indices.begin(), agent_indices.end(), _rng);
+
   // Process actions by priority levels (highest to lowest)
   for (unsigned char offset = 0; offset <= _max_action_priority; offset++) {
     unsigned char current_priority = _max_action_priority - offset;
 
-    for (size_t agent_idx = 0; agent_idx < _agents.size(); agent_idx++) {
+    for (const auto& agent_idx : agent_indices) {
```
Comment (sasmith) [inline mettagrid/src/metta/mettagrid/mettagrid_c.cpp:365]: Any reason to prefer this vs the previous version? I expect the answer is "not really", but curious if there's something to learn here.

**Review by sasmith — APPROVED** [review]
Fix the workspace file
